### DISCUSSION TITLE
How can I have access to the Jupyter API from a different Vscode extension?

### DISCUSSION THREAD

**Discussion by micoloth** [microsoft/vscode-jupyter#13331] [General]
Hello!

## TL;DR: 

I'd like to write a new Vscode Extension to enable new functionalities (code analysis, etc) for a Python environment running in Vscode.

In order to do this, I'd like my extension to access the `vscode-jupyter` api. 

But when I try, i get this message:

```
Thanks for trying the Jupyter API. Please file an issue on our repo to use this API in production. This would prevent us from breaking your extension when updating the API (as it is still a work in progress).
```

Is there a way to use the API, or is the access restricted?

## Details:

In order to develop new functionalities for Python in Vscode, I'd like my extension to connect to a running Python Kernel _managed by `vscode-jupyter`._

> **Why:** The idea is, I'd like to let `vscode-jupyter` handle all the complicated logic of spinning up the kernel, open the Interactive Window panel, handle the Kernel Picker menus, etc. Then, connect to the same running kernel and run my logic in it. 
> 
> In this way, there is only 1 Python instance running, and I can analyse what the user is doing (in the same way the `VariableViewer` panels injects python code in the running kernel).

After a bit of digging, I've found out that this possibility exists: in particular, the extension `vscode-jupyter-powertoys` does that, by calling the Jupyter API in this way:
```
  const jupyterExt = extensions.getExtension<JupyterAPI>('ms-toolsai.jupyter');
  await jupyterExt.activate();
  const kernelService = await jupyterExt.exports.getKernelService();
  const activeKernels = kernelService.getActiveKernels();
```

Unfortunately, i see that the access to this feature is restricted, with this code granting access to Trusted developers only:

```
  if (!TrustedExtensionPublishers.has(publisherId) || PublishersAllowedWithPrompts.has(publisherId)) {
      this.appShell
          .showInformationMessage(DataScience.thanksForUsingJupyterKernelApiPleaseRegisterWithUs)
```

which prints this nice message 

```
  'Thanks for trying the Jupyter API. Please file an issue on our repo to use this API in production. This would prevent us from breaking your extension when updating the API (as it is still a work in progress).'
```

So I'm opening this issue here.

My question is, is there a way to be enabled to use this experimental feature? Or is it restricted to the Microsoft team only?

How can I have access to the API?

Thank you in advance!

**Comment by DonJayamanne** [comment]
Thanks for filing this issue and expressing interest in using the API.

> My question is, is there a way to be enabled to use this experimental feature? Or is it restricted to the Microsoft team only?

Currently this approach is added so we have control over what API is used by which extension.
E.g. if an API changes, then we need to bear the maintenance cost of updating the extensions using this API without breaking them, and the like.

**A few questions/notes**
* Will your extension be open source or closed source
* Would you be able to provide more information about how this will be used (please feel free to send me an email if you'd like to discuss this offline) (you can reach me on  `don` + '.' + 'jayamanne' and company i work for_

**Reply by micoloth** [reply]
Hi DonJayamanne.

Sure! I'll write you shortly.

Thanks for answering

**Reply by ehassaan** [reply]
@DonJayamanne ,  I have a similar use-case. My extension currently provides SQL notebook using DuckDb. But I want to add support for spark code as well. For that I need to find kernels and execute code against them. It would be great if I can rely on some of the functionality already implemented by this extension.

https://marketplace.visualstudio.com/items?itemName=ducklake.ducklab

**Reply by micoloth** [reply]
Good news @HassaanAkbar . This exists now!

They finally release the API (I guess the end goal is allow Copilot to mess with it, correct me if I'm wrong @DonJayamanne 😬)

It was months ago I don't remember the details. But if you are interested you can have a look at the extension that I wrote [here](https://github.com/micoloth/vscode-reactive-jupyter/blob/main/src/extension.ts)  (I made it for my personal use, no stress on the code style. But it mostly works)

**Reply by ehassaan** [reply]
Looks cool. I will take a look at it. Thanks a lot

**Reply by ehassaan** [reply]
Looks like this API is not sufficient for my use-case. I want kernel picker in my .isql notebook to show similar options as shown for .ipynb notebook then I need to launch kernels from a python environment and execute code against it. I don't want to require user to create a separate notebook just for starting the kernel.

**Reply by yiliang114** [reply]
So is it possible to expose the api to web extension for use?
